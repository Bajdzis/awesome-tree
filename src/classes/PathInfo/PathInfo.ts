import { splitStringWithSplitter } from '../../utils/splitStringWithSplitter/splitStringWithSplitter';
import { WordsInfo } from '../WordsInfo/WordsInfo';



export class PathInfo {

    private path: string;
    private parts: WordsInfo[];
    private type: 'file' | 'directory';

    constructor(path:string) {
        this.path = path;
        this.type = path.match(/[\\/]+$/i) ? 'directory': 'file';
        this.parts = splitStringWithSplitter(path,'\\/.')
            .filter(part => part.match(/([a-z]+([-|_]{1,1}[a-z]*)*)+/i))
            .map(part => new WordsInfo(part));
    }

    isFile() {
        return this.type === 'file';
    }

    isDirectory() {
        return this.type === 'directory';
    }

    getExtension(){
        if(this.isDirectory()){
            return '/DIRECTORY/';
        }
        return (/.([a-z]+)$/i).exec(this.path)?.[1];
    }

    getParent() {
        if (this.isFile()) {
            const directory = this.path.replace(/[^/\\]+$/, '');
            return new PathInfo(directory);
        }
        const parentDirectory = this.path.replace(/[^/\\]+(\/|\\)$/, '');
        return new PathInfo(parentDirectory);
    }

    getName(){
        const result = this.path.match( this.isFile() ?  /([^/\\]+)$/ : /([^/\\]+)[/\\]{1,1}$/);
        if(result) {
            return result[1];
        }

        return '';
    }

    getPath() {
        return this.path;
    }

    getParts() {
        return [...this.parts];
    }

    isSameNumberOfParts(instanceToCompare: PathInfo){
        return instanceToCompare.getParts().length === this.getParts().length;
    }

    isSimilarWords(instanceToCompare: PathInfo){
        if(!this.isSameNumberOfParts(instanceToCompare)) {
            return false;
        }
        const partsToCompare = instanceToCompare.getParts();
        return this.parts.every((part, index) => part.isSimilarly(partsToCompare[index]));
    }

    isSimilar(instanceToCompare: PathInfo){
        return this.isSimilarWords(instanceToCompare) && this.getExtension() === instanceToCompare.getExtension();
    }

    getOnlySimilarPath(paths: PathInfo[]){
        return paths.filter(path => this.isSimilar(path));
    }

    includes(partPath: PathInfo) {
        if(partPath.isFile() || partPath.getParts().length > this.getParts().length){
            return false;
        }
        return partPath.getParts().every((part,index) => part.isSame(this.getParts()[index]));
    }

    includesSimilarly(partPath: PathInfo) {
        if(partPath.isFile() || partPath.getParts().length > this.getParts().length){
            return false;
        }
        return partPath.getParts().every((part,index) => part.isSimilarly(this.getParts()[index]));
    }
}
