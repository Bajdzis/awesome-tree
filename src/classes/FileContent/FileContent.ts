import { FileContentNode } from '../FileContentNode/FileContentNode';
import { PathInfo } from '../PathInfo/PathInfo';

interface LineInfo {
    line: string;
    whiteSpaces:  number;
}
export class FileContent {
    private path: PathInfo;
    private content: string;

    constructor(path: PathInfo, content: string) {
        if(path.isDirectory()){
            throw new Error('[FileContent] Pass directory not file !');
        }
        this.path = path;
        this.content = content;
    }

    getContent(){
        return this.content;
    }

    getPath(){
        return this.path;
    }

    getExtension(){
        return this.path.getExtension();
    }

    getFileGraph(): FileContentNode {
        const file = new FileContentNode();
        const lines = this.content.split('\n');

        const countWhitespace = lines.map<LineInfo>((line, index) => {

            const extendString = index === lines.length -1 ? line : `${line}\n`;

            if (line.trim().length === 0) {
                return {
                    line: extendString,
                    whiteSpaces: Math.max(
                        FileContent.countNumberOfWhitespace(lines?.[index+1] || ''),
                        FileContent.countNumberOfWhitespace(lines?.[index-1] || '')
                    )
                };
            }

            return {
                line: extendString,
                whiteSpaces: FileContent.countNumberOfWhitespace(line)
            };
        });

        const children = this.nodeLine(countWhitespace);

        return file.setChildren(children);
    }

    nodeLine(lines:LineInfo[]): FileContentNode[]{
        if (lines.length === 0 ) {
            return [];
        }
        if (lines.length === 1 ) {
            return [new FileContentNode(lines[0].line)];
        }
        const baseWhitespace = lines[0].whiteSpaces;
        let lastIndexWithBaseSpace = 0;
        let omitElementsFromEnd = 0;
        const result: FileContentNode[] = [];

        for (let i = 1; i < lines.length; i++) {
            const prevLine = lines[i-1];
            const line = lines[i];


            if(line.whiteSpaces === baseWhitespace) {
                if(prevLine.whiteSpaces === line.whiteSpaces) {
                    const node = new FileContentNode(prevLine.line);
                    result.push(node);
                    omitElementsFromEnd = 1;
                } else {
                    const start = lines[lastIndexWithBaseSpace];
                    const node = new FileContentNode(start.line, line.line);
                    const childrenNodes = this.nodeLine(lines.slice(lastIndexWithBaseSpace + 1, i));
                    result.push(node.setChildren(childrenNodes));
                    omitElementsFromEnd = 1;
                    i++;
                }
                lastIndexWithBaseSpace = i;
            } else {
                omitElementsFromEnd++;
            }
        }

        if(omitElementsFromEnd!==0) {
            const omitLines = lines.slice(lines.length - omitElementsFromEnd, lines.length);
            omitLines.forEach(line => {
                result.push(new FileContentNode(line.line));
            });
        }

        return result;
    }

    static countNumberOfWhitespace(s0: string){
        for (let i = 0; i < s0.length; i++) {
            const letter = s0[i];
            if(letter === '\t' || letter === ' ') {
                continue;
            }
            return i;
        }

        return 0;
    }

}
