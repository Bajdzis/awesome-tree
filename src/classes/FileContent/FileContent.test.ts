import { PathInfo } from '../PathInfo/PathInfo';
import { FileContent } from './FileContent';

const content = `
class Awesome {

    doAwesomeStuff() {
        // TODO: implement
        // TODO: this
        // TODO: function
    }

    doOtherAwesomeStuff() {
        // TODO: implement!
        return true;
    }

}
`;

describe('FileContent', () => {

    it('should get info about file and path' , () => {
        const path = new PathInfo('/home/awesomeClass.ts');
        const fileContent = new FileContent(path, content);

        expect(fileContent.getExtension()).toEqual('ts');
        expect(fileContent.getContent()).toEqual(content);
    });

    it('should generated correct nodes base on space number' , () => {
        const path = new PathInfo('/home/awesomeClass.ts');
        const fileContent = new FileContent(path, content);

        const result = fileContent.getFileGraph();

        expect(result.getContent()).toEqual(content);
        expect(result).toMatchSnapshot();
    });
});



