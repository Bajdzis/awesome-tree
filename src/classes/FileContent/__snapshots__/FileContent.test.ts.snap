// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileContent should generated correct nodes base on space number 1`] = `
FileContentNode {
  "body": Array [
    FileContentNode {
      "body": Array [],
      "end": "",
      "start": "
",
    },
    FileContentNode {
      "body": Array [
        FileContentNode {
          "body": Array [],
          "end": "",
          "start": "
",
        },
        FileContentNode {
          "body": Array [
            FileContentNode {
              "body": Array [],
              "end": "",
              "start": "        // TODO: implement
",
            },
            FileContentNode {
              "body": Array [],
              "end": "",
              "start": "        // TODO: this
",
            },
            FileContentNode {
              "body": Array [],
              "end": "",
              "start": "        // TODO: function
",
            },
          ],
          "end": "    }
",
          "start": "    doAwesomeStuff() {
",
        },
        FileContentNode {
          "body": Array [],
          "end": "",
          "start": "
",
        },
        FileContentNode {
          "body": Array [
            FileContentNode {
              "body": Array [],
              "end": "",
              "start": "        // TODO: implement!
",
            },
            FileContentNode {
              "body": Array [],
              "end": "",
              "start": "        return true;
",
            },
          ],
          "end": "    }
",
          "start": "    doOtherAwesomeStuff() {
",
        },
        FileContentNode {
          "body": Array [],
          "end": "",
          "start": "
",
        },
      ],
      "end": "}
",
      "start": "class Awesome {
",
    },
    FileContentNode {
      "body": Array [],
      "end": "",
      "start": "",
    },
  ],
  "end": "",
  "start": "",
}
`;
